ledger/ledger-api-test-tool/src/main/scala/com/daml/ledger/api/testtool/suites/CommandDeduplicationPeriodValidationIT.scala

```diff
@@ -161,30 +162,76 @@ class CommandDeduplicationPeriodValidationIT extends LedgerTestSuite {
   )(implicit ec => { case Participants(Participant(ledger, party)) =>
+    def submitAndWaitWithDeduplication(
+        deduplicationPeriod: DeduplicationPeriod.DeduplicationOffset
+    ) = {
+      ledger
+        .submitAndWait(
+          ledger
+            .submitAndWaitRequest(party, Dummy(party).create.command)
+            .update(
+              _.commands.deduplicationPeriod := deduplicationPeriod
+            )
+        )
+    }
+    val isOffsetNativlySupported =
+      ledger.features.commandDeduplicationFeatures.getDeduplicationPeriodSupport.offsetSupport.isOffsetNativeSupport
     for {
+      start <- ledger.currentEnd()
       firstCreate <- ledger.create(party, Dummy(party))
-      firstExercise <- ledger.exercise(party, firstCreate.exerciseDummyChoice1)
-      end <- ledger.currentEnd()
+      _ <- ledger.exercise(party, firstCreate.exerciseDummyChoice1)
       secondCreate <- ledger.create(party, Dummy(party))
+      _ <- ledger.submitAndWait(ledger.submitAndWaitRequest(party, Dummy(party).create.command))
+      end <- ledger.currentEnd()
       _ <- ledger.exercise(party, secondCreate.exerciseDummyChoice1)
       _ <- ledger.create(party, Dummy(party)) // move ledger end
+      _ <- ledger.submitAndWait(ledger.submitAndWaitRequest(party, Dummy(party).create.command))
       _ <- ledger.prune(pruneUpTo = end)
-      deduplicationPeriod = DeduplicationPeriod.DeduplicationOffset(
-        Ref.HexString.assertFromString(firstExercise.offset)
-      )
-      _ <- assertFailedRequest(
+      failure <- submitAndWaitWithDeduplication(
+        DeduplicationPeriod.DeduplicationOffset(
+          start.getAbsolute
+        )
+      ).mustFail("using an offset which was pruned")
+      _ = assertGrpcErrorRegex(
         ledger,
-        party,
-        deduplicationPeriod,
-        failReason = "Submitting a command with a pruned offset",
-        expectedMessage = ".*",
-        expectedCode = Status.Code.INVALID_ARGUMENT,
-        expectedError = LedgerApiErrors.RequestValidation.ParticipantPrunedDataAccessed,
+        failure,
+        expectedCode = Status.Code.FAILED_PRECONDITION,
+        // Canton returns INVALID_DEDUPLICATION_PERIOD with earliest_offset metadata
+        // KV returns PARTICIPANT_PRUNED_DATA_ACCESSED with earliest_offset metadata
+        selfServiceErrorCode =
+          if (isOffsetNativlySupported)
+            LedgerApiErrors.RequestValidation.InvalidDeduplicationPeriodField
+          else LedgerApiErrors.RequestValidation.ParticipantPrunedDataAccessed,
+        None,
       )
+      earliestOffset = extractErrorInfoMetadataValue(failure, "earliest_offset")
+      _ <-
+        // Because KV treats deduplication offsets as inclusive, and because the participant pruning offset is inclusive
+        // we cannot simply use the received offset as a deduplication period, but we have to find the first completion after the given offset
+        if (isOffsetNativlySupported) {
```

Comment:
Maybe pull the part that finds the earliest offset out into its own function for clarity. (I think the call to `submitAndWaitWithDeduplication` should stay in the test definition.)

Reply:
Done, good point